Improve mount and format packages

dave-tucker opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/linuxkit/linuxkit/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
- `format`
  - Added supports writing partition labels
  - Added support for creating `btrfs` and `xfs` partitions
- `mount`
  - Allowed for a disk to be specified by name, label or UUID
  - Fixed automatic mount logic to take already mounted disks and partitions in to account.
- Added `extend` package to extend partitions on disks
- Extending disks with the new `extend` package
- Added test cases and updated documentation

**- How I did it**

Typing.

**- How to verify it**

Run ze tests!

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Add multiple disk/mountpoint support for format and mount packages

**- A picture of a cute animal (not mandatory but encouraged)**
![](https://images.duckduckgo.com/iu/?u=http%3A%2F%2Fwww.indiaonrent.com%2Fforwards%2Fm%2Fmultiple-animals-being-cute-together%2Fres%2Fks2h0f.jpg&f=1)

justincormack commented on pkg/mount/README.md:
```diff
@@ -0,0 +1,34 @@
+
+Mount block devices for use in Linuxkit
+
+**NOTE: At this time, block devices may only be mounted in `/var`**
```
why?

dave-tucker replied:
Because it confused me endlessly when my mount to `/boot` kept failing.
It seems that because only `/dev` and `/var` are bound by default in the `mount` container mounts outside of those locations will never propagate to the host.

dave-tucker replied:
I can caveat this with `unless you add the location to the binds: section`

justincormack replied:
Yes it would be better to document this. We could also add some more standard mount points eg `/mnt`.

dave-tucker commented:
This is now complete, fixed, tested and ready for review 🎉

justincormack commented on pkg/extend/extend.go:
```diff
@@ -0,0 +1,229 @@
+				return nil
+			case 2, 3:
+				log.Printf("fsck fixed errors but requires a reboot. rebooting")
+				exec.Command("reboot").Run()
```
Not sure we should reboot really. Actually not sure when this might happen either.

justincormack replied:
Also you dont have capabilities to perform a reboot...

justincormack commented on pkg/format/README.md:
```diff
@@ -0,0 +1,31 @@
+onboot:
+  name: format
+  image: linuxkit/format:${hash}
+  command: ["/usr/bin/format", "/dev/vda", "/dev/vdb" ]
```
We should not support multiple formats in the same command. We will need additional parameters, eg filesystem type.

rn replied:
Could we also add this to `docs/external-disk.md`

justincormack commented on pkg/mount/README.md:
```diff
@@ -0,0 +1,34 @@
+onboot:
+  name: format
+  image: linuxkit/format:${hash}
+  command: ["/usr/bin/format", "/var/lib/docker:/dev/vda1", "/var/foo:/dev/vdb1" ]
```
that should be `mountie` not `format`. Again, we dont really want to support multiple ones as per #2173

deitch replied:
@dave-tucker I liked this identical format in #2173 but consensus was one format per container. I think the thrust was that having multiple `onboot` containers, each with the same image but different options to format or mount a different disk, is a near-zero-cost overhead. 

I think my own leaning was traditional. We do run `mount` once per disk/mountpoint, which would be equivalent to one container. But when we have all of the mount points in a config (`fstab`), a single command mounts them all, which is the equivalent of a single container.

riyazdf commented on pkg/extend/extend.go:
```diff
@@ -0,0 +1,229 @@
+				return nil
+			case 2, 3:
+				log.Printf("fsck fixed errors but requires a reboot. rebooting")
+				exec.Command("reboot").Run()
```
same as @justincormack's reboot comment above

riyazdf commented on pkg/extend/extend.go:
```diff
@@ -0,0 +1,229 @@
+	devs, _ := ioutil.ReadDir("/dev")
+	for _, d := range devs {
+		// this probably shouldn't be so hard
+		// but d.Mode()&os.ModeDevice == 0 doesn't work as expecte
```
super nit: s/expecte/expected

riyazdf commented on pkg/extend/README.md:
```diff
@@ -0,0 +1,30 @@
+extend
```
these READMEs are awesome, thank you!

rn replied:
I agree, but could we add this to `docs/external-disk.md` instead?

riyazdf commented on pkg/format/format.go:
```diff
@@ -0,0 +1,158 @@
+	devs, _ := ioutil.ReadDir("/dev")
+	for _, d := range devs {
+		// this probably shouldn't be so hard
+		// but d.Mode()&os.ModeDevice == 0 doesn't work as expecte
```
same nit: s/expecte/expected

riyazdf commented on pkg/mount/mountie.go:
```diff
@@ -0,0 +1,152 @@
+	devs, _ := ioutil.ReadDir("/dev")
+	for _, d := range devs {
+		// this probably shouldn't be so hard
+		// but d.Mode()&os.ModeDevice == 0 doesn't work as expecte
```
same typo nit

riyazdf commented on test/cases/040_packages/009_extend/test-create.yml:
```diff
@@ -0,0 +1,26 @@
+    image: "linuxkit/poweroff:bce51402e293da0b653923a43c3c7be6e0effa05"
+    command: ["/bin/sh", "/poweroff.sh", "10"]
+trust:
+#  org: [linuxkit]
```
todo: uncomment after sign&push

justincormack replied:
may as well uncomment now, as the tests not passing anyway until then

deitch commented:
Curious: `mount.sh` was quite small. What was the benefit of replacing it with a compiled go binary? I imagine the performance benefits on startup were tiny, and that it wasn't difficult to parse the options with standard tools. Was it just to be able to have a single-binary container? Or some other reason?

justincormack commented:
Its not a single binary container yet, as it still calls `fdisk` etc. But it has better error checking and is easier to extend (there are more complexities with other types of disks).

justincormack commented:
@dave-tucker needs a rebase

dave-tucker commented:
Rebased and updated based on review comments so far.
I've incorporated the features from #2173 and added tests for them.
I added support for writing partition labels in `format` and also allowed creation of `xfs` and `btrfs` partitions.

Finally I added a new tests for the above ✅ 

Note: This will fail CI because the images need pushing/signing

dave-tucker commented:
Side Note: I've got code written for writing partition tables in Go without the need for `sfdisk`.
`mount` can be replaced with a `syscall`. The fanciness of mounting by `uuid` or `label` would require either `udev` to populate `/dev/disk/by-label` or some `cgo` bindings to `libblkid`. The only dependency would then would be the `mkfs.*` utilities which could also be replaced with `cgo` if it were required.

justincormack commented:
@dave-tucker for anything non trivial like partition tables (of which there are for example two or three sorts we need to implement) we should create a new repo, with extensive (LinuxKit based!) tests and use code from there.

ijc commented:
What is the motivation for spending the effort to reimplement something subtle and tricksie like partition table creation (or mkfs for that matter) when there are perfectly good and scriptable tools already in existence? Just having it in Go doesn't seem to justify the effort to me.

dave-tucker commented:
> What is the motivation for spending the effort to reimplement something subtle and tricksie like partition table creation (or mkfs for that matter) when there are perfectly good and scriptable tools already in existence?

To be clear, I had some free time last night and decided to have a crack at the code for MBR partition tables just for fun. I then took a quick look to see if it were feasible to replace everything with Go out of curiosity.

There is no motivation beyond curiosity for me at this time and I've not committed to spending any time on it.

justincormack commented on docs/external-disk.md:
````diff
@@ -19,45 +19,108 @@ The `-disk` specification may be repeated for multiple disks, although a limited
+```
+onboot:
+  name: mount
+  image: linuxkit/format:${hash}
````
this should be `linuxkit/mount`

justincormack commented on docs/external-disk.md:
````diff
@@ -19,45 +19,108 @@ The `-disk` specification may be repeated for multiple disks, although a limited
+```
+onboot:
+  name: format
+  image: linuxkit/format:${hash}
````
and here

justincormack commented on docs/external-disk.md:
```diff
@@ -19,45 +19,108 @@ The `-disk` specification may be repeated for multiple disks, although a limited
+onboot:
+  name: extend
+  image: linuxkit/extend:${hash}
+  command: ["/usr/bin/extend", "/dev/vda"]
```
It kind of seems more natural to specify `/dev/vda1` to me as that is what is extended though its a bit weird either way...

dave-tucker replied:
agree that either way is weird. I just went with the physical device name as it's what `sfdisk` uses to read the partition table and grab the name of the first partition. I didn't fancy guessing the correct device name from something like `/dev/nvme0n1p1` 😁

justincormack commented on test/cases/040_packages/009_extend/test-create.yml:
```diff
@@ -0,0 +1,26 @@
+trust:
+#  org: [linuxkit]
+  image:
+    - alpine:3.6
```
should be `org: library`

dave-tucker commented:
Review comments addressed.

dave-tucker commented:
If someone is kind enough to push/sign the `pkg/format`, `pkg/mount` and `pkg/extend` images then hopefully CI should be green

riyazdf commented:
signed and pushed:
- `linuxkit/extend:67da9668a0561fd97a17e4d17fbd573a992ef86c`
- `linuxkit/format:64582e067d4b042e6f304cb679912f863d588b87`
- `linuxkit/mount:7517c010005ee48a2ac107aa1b91a7f978272d0e`

Kicking CI now.

riyazdf commented:
@dave-tucker hmm, looks like a failure with the format_mount.mulitiple test case?
 
https://github.com/linuxkit/linuxkit-logs/blob/linuxkit-build-of-a367f7efd4b27f4b2512b6cd7c16d1c15330a6cb-default-ci-pr/value/aaf8e019-5a36-49a6-918e-15b779236911/linuxkit.packages.format_mount.multiple.log

dave-tucker commented:
@riyaz thanks! I tested with hyperkit where devices where /dev/vda. Looks like they are different on qemu. Will fix the tests this morning!

dave-tucker commented:
yay! looks like tests are green. will rebase now.

justincormack commented:
needs a rebase

dave-tucker commented:
rebased but the tests hang due to the moby/containerd changes 😢 
fixing the tests and force pushing momentarily

dave-tucker commented:
Huzzah! Rebased. All Green. Ready to go!

dave-tucker commented:
rebased.

dave-tucker commented:
@deitch @rn @riyazdf @justincormack please could I have some reviews?

rn commented on pkg/mount/Dockerfile:
```diff
@@ -1,20 +1,27 @@
     && true
 RUN rm -rf /out/etc/apk /out/lib/apk /out/var/cache
 
+FROM linuxkit/alpine:9bcf61f605ef0ce36cc94d59b8eac307862de6e1 AS mirror
```
see above

justincormack commented on pkg/extend/extend.go:
```diff
@@ -0,0 +1,227 @@
+		return fmt.Errorf("fsck before resize failed: %v", err)
+	}
+
+	if err := exec.Command("resize2fs", partition.Node).Run(); err != nil {
```
Can we add a filesystem type argument to this too to match the others, even if we just error out if not `ext4` to start with?

rn commented on test/cases/040_packages/008_format_mount/000_auto/test.sh:
```diff
@@ -0,0 +1,26 @@
+
+# Test code goes here
+moby build -name ${NAME} -output kernel+initrd test.yml
+RESULT="$(linuxkit run qemu -disk file=${NAME}1.img,size=512M ${NAME})"
```
is there a reason this and the other tests for the use of `qemu`?

dave-tucker replied:
yes. the device name of drives varies from backend to backend e.g `/dev/vda1` on Hyperkit and `/dev/sda1` on qemu. As some of theses tests require a device name it made sense to me to pin the backend to qemu as it's available on all platforms.

rn replied:
ah, how annoying

rn replied:
actually, this test does not rely on the name of the device, so shouldn't really hard code `qemu`

dave-tucker replied:
this one doesn't, but the `by_name` test does. I think I have an idea to avoid pinning to `qemu`. will try it out and re-push

rn commented on test/cases/040_packages/008_format_mount/000_auto/test.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+# SUMMARY: Check that the format and mount packages work
```
could the summaries of the tests be more specific?

deitch commented on docs/external-disk.md:
```diff
@@ -19,45 +19,108 @@ The `-disk` specification may be repeated for multiple disks, although a limited
-Notice several key points:
+`-type` can be used to specify the type e.g ext4
+`-label` can be used to give the disk a label
+The final (optional) argument specifies the device name
```
Does this also fill the entire disk like above? I assume that any more complex partition tables are not doable here, should either be done outside or once in LinuxKit running? Probably should add to the docs.

deitch commented on docs/external-disk.md:
````diff
@@ -19,45 +19,108 @@ The `-disk` specification may be repeated for multiple disks, although a limited
+  name: mount
+  image: linuxkit/mount:${hash}
+  command: ["/usr/bin/mountie", "-device", "/dev/sda1", "/var/lib/docker" ]
+```
````
I know it seems painfully obvious, but if someone wants to format or mount multiple disks, they should just run the `format` container or `mount` container multiple times, correct? We probably should say so and give examples. You _know_ that someone will ask on the channel (or even submit a PR) with all sorts of syntaxes for formatting or mounting multiple disks at once. We even tried to do it here. Head that one off.

deitch commented on docs/external-disk.md:
````diff
@@ -19,45 +19,108 @@ The `-disk` specification may be repeated for multiple disks, although a limited
+```
+onboot:
+  name: extend
+  image: linuxkit/extend:${hash}
````
Same comment @rn had about `${hash}` vs `<hash>`

deitch replied:
Same comment @rn had about `${hash}` vs `<hash>`

dave-tucker commented:
@deitch @rn rebased and review comments addressed.
@justincormack I added support for extending `xfs` and `btrfs` partitions
This will fail CI (again) due to the SHA's changing when I updated the Dockerfiles.
If someone would be kind enough to push and sign for me that'd be wonderful!

rn commented on docs/external-disk.md:
```diff
@@ -19,45 +19,112 @@ The `-disk` specification may be repeated for multiple disks, although a limited
+
+**NOTE: Block devices may only be mounted in `/var` unless you have explicitly added an additional bind mount**
+
+In this mode of operation, the first unmounted linux partition on the any block device is mounted to the mountpoint provided.
```
'on the any' -> "on the first"?

rn replied:
"In this mode"..it doesn't say that there are multiple modes for mounting...

Maybe:
The simplest mode is ...

rn commented on docs/external-disk.md:
````diff
@@ -19,45 +19,112 @@ The `-disk` specification may be repeated for multiple disks, although a limited
-2. `mount` which mounts the filesystem to a provided path
+```
+onboot:
+  name: format
````
the syntax here and for all the examples is wrong: it's missing the `-` in front of name

dave-tucker replied:
ack

rn commented on pkg/extend/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+    && true
+RUN rm -rf /out/etc/apk /out/lib/apk /out/var/cache
+
+FROM linuxkit/alpine:9bcf61f605ef0ce36cc94d59b8eac307862de6e1 AS build
```
any reason you are using two different alpine base images in the same Dockerfile?

rn commented on pkg/format/Dockerfile:
```diff
@@ -6,17 +6,27 @@ RUN apk add --no-cache --initdb -p /out \
     && true
 RUN rm -rf /out/etc/apk /out/lib/apk /out/var/cache
 
+FROM linuxkit/alpine:9bcf61f605ef0ce36cc94d59b8eac307862de6e1 AS build
```
And this uses a different alpine base to the other Dockerfiles. It would be good if a PR updates them somewhere that it would update the base in the other dockerfiles it touches as well

rn commented on pkg/mount/Dockerfile:
```diff
@@ -4,17 +4,24 @@ RUN mkdir -p /out/etc/apk && cp -r /etc/apk/* /out/etc/apk/
     && true
 RUN rm -rf /out/etc/apk /out/lib/apk /out/var/cache
 
+FROM linuxkit/alpine:9bcf61f605ef0ce36cc94d59b8eac307862de6e1 AS build
```
ditto

rn commented on test/cases/040_packages/008_format_mount/003_btrfs/test.sh:
```diff
@@ -0,0 +1,26 @@
+
+# Test code goes here
+moby build -name ${NAME} -output kernel+initrd test.yml
+RESULT="$(linuxkit run qemu -disk file=${NAME}1.img,size=512M ${NAME})"
```
this test also does not rely on the device name, so `qemu` can be removed

rn commented on test/cases/040_packages/008_format_mount/004_xfs/test.sh:
```diff
@@ -0,0 +1,26 @@
+
+# Test code goes here
+moby build -name ${NAME} -output kernel+initrd test.yml
+RESULT="$(linuxkit run qemu -disk file=${NAME}1.img,size=512M ${NAME})"
```
this one also does not rely on the device name

dave-tucker commented:
@rn comments addressed. None of the tests are hardcoded with qemu now. To fix the `by_name` test I checked in a `test.yml.orig` and used `sed` to inject the correct device name based on `RT_OS`. For the `multiple` test I removed the device names as they are optional anyway.

rn commented on test/cases/040_packages/008_format_mount/001_by_label/test.yml:
```diff
@@ -0,0 +1,30 @@
+onboot:
+  - name: format
+    image: linuxkit/format:efafddf9bc6165b5efaf09c532c15a1100a10e61
+    command: ["/usr/bin/format", "-label", "docker", "/dev/sda"]
```
This still hard codes the `/dev/sda` so presumably fails when executed on hyperkit

rn commented on test/cases/040_packages/008_format_mount/002_by_name/test.sh:
```diff
@@ -0,0 +1,32 @@
+	DEVICE="/dev/sda"
+fi
+
+sed -e "s,@DEVICE@,${DEVICE},g" test.yml.orig > test.yml
```
doesn't `test.yml` needs cleaning up after the test?

rn commented on test/cases/040_packages/009_extend/000_ext4/test.sh:
```diff
@@ -0,0 +1,35 @@
+DISK=disk0.img
+clean_up() {
+	find . -depth -iname "${NAME}*" -not -iname "*.yml" -exec rm -rf {} \;
+#	rm -rf ${DISK} || true
```
why is this commented out? If it's not needed, remove it

dave-tucker commented:
@rn nits addressed and rebased.

dave-tucker commented:
Could someone (@rn @justincormack @deitch) sign and push these images for me?
It'd be nice to make CI green!

Note: I'm on vacation from Thurs for a week and a bit so would be nice for this to merge before I go...

rn commented:
Pushed/signed:
- `linuxkit/extend:1e81ffe40ad63887d6210228c2a791f28375ee0f`
- `linuxkit/format:efafddf9bc6165b5efaf09c532c15a1100a10e61`
- `linuxkit/mount:54990a6a69cb3ead4da8a9c1f0b651e27aea8d3f`

You need to update the hashes in for `linuxkit/init` and `linuxkit/runc` in the test YAML files. An update should trigger CI to run.

dave-tucker commented:
Done. CI's running... fingers crossed 🙏

dave-tucker commented:
Huzzah! It's all green.
The only thing I haven't done here is update the samples.
I'd be happy to add commits to do that, or, do so in a follow up PR as there is already quite a lot in this one...

rn commented:
could you add a commit to update the examples? Might as well do everything in one PR

dave-tucker commented:
Done

justincormack commented:
If you have a moment can you remove the `page-poison=1` from the test yaml files, and any image names in quotes, as they all got cleaned up before and reintroducing inconsistency is annoying, will save another PR to remove. Otherwise looks ok

dave-tucker commented:
@justincormack done

dave-tucker commented:
@rn done! waiting on CI now...
